iib/web/models.py

```diff
@@ -2200,3 +2214,159 @@ def get_mutable_keys(self) -> Set[str]:
+        return '<Deprecation schema={0!r}>'.format(self.schema)
+
+    @classmethod
+    def get_or_create(cls, deprecation_schema: str) -> DeprecationSchema:
+        """
+        Get the deprecation schema from the database and create it if it doesn't exist.
+
+        :param str deprecation_schema: the deprecation schema for an operator
+        :return: a DeprecationSchema object based on the input name; the DeprecationSchema object
+            will be added to the database session, but not committed, if it was created
+        :rtype: DeprecationSchema
+        """
+        # cls.query triggers an auto-flush of the session by default. So if there are
+        # multiple requests with same parameters submitted to IIB, call to query pre-maturely
+        # flushes the contents of the session not allowing our handlers to resolve conflicts.
+        # https://docs.sqlalchemy.org/en/20/orm/session_api.html#sqlalchemy.orm.Session.params.autoflush
+        with db.session.no_autoflush:
+            schema = cls.query.filter_by(schema=deprecation_schema).first()
+        if not schema:
+            schema = DeprecationSchema(schema=deprecation_schema)
+            try:
+                # This is a SAVEPOINT so that the rest of the session is not rolled back when
+                # adding the image conflicts with an already existing row added by another request
+                # with similar pullspecs is submitted at the same time. When the context manager
+                # completes, the objects local to it are committed. If an error is raised, it
+                # rolls back objects local to it while keeping the parent session unaffected.
+                # https://docs.sqlalchemy.org/en/20/orm/session_transaction.html#using-savepoint
+                with db.session.begin_nested():
+                    db.session.add(schema)
+            except sqlalchemy.exc.IntegrityError:
+                current_app.logger.info('Schema is already in database. "%s"', deprecation_schema)
+            schema = cls.query.filter_by(schema=deprecation_schema).first()
+
+        return schema
+
+
+class RequestAddDeprecations(Request, RequestIndexImageMixin):
+    """An "add-deprecations" index image build request."""
+
+    __tablename__ = 'request_add_deprecations'
+
+    id: Mapped[int] = db.mapped_column(
+        db.ForeignKey('request.id'), autoincrement=False, primary_key=True
+    )
+    operator_package: Mapped[Operator] = db.relationship(
+        'Operator', secondary=RequestAddDeprecationsOperator.__table__
+    )
+    deprecation_schema: Mapped[DeprecationSchema] = db.relationship(
+        'DeprecationSchema', secondary=RequestAddDeprecationsDeprecationSchema.__table__
+    )
+
+    __mapper_args__ = {
+        'polymorphic_identity': RequestTypeMapping.__members__['add_deprecations'].value
+    }
+
+    @classmethod
+    def from_json(  # type: ignore[override] # noqa: F821
+        cls,
+        kwargs: AddDeprecationRequestPayload,
```

Comment:
IMO using `kwargs` as an argument instead of a list of kw arguments `**kwargs` is a bit weird... What about naming it `payload` or something simlilar?

Reply:
Ok, agreed!

Marking it as resolved